File: qc-domain/src/main/java/com/eager/questioncloud/domain/social/GoogleAPI.java
package com.eager.questioncloud.domain.social;

import com.eager.questioncloud.domain.user.vo.AccountType;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.BodyInserters;
import org.springframework.web.reactive.function.client.WebClient;

@Component
public class GoogleAPI extends SocialAPI {
    @Value("${GOOGLE_CLIENT_ID}")
    private String GOOGLE_CLIENT_ID;

    @Value("${GOOGLE_CLIENT_SECRET}")
    private String GOOGLE_CLIENT_SECRET;

    @Value("${CLIENT_URL}")
    private String CLIENT_URL;

    @Override
    public String getAccessToken(String code) {
        WebClient webClient = WebClient.create("https://oauth2.googleapis.com/token");
        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();

        formData.add("client_id", GOOGLE_CLIENT_ID);
        formData.add("client_secret", GOOGLE_CLIENT_SECRET);
        formData.add("code", code);
        formData.add("grant_type", "authorization_code");
        formData.add("redirect_uri", CLIENT_URL + "/user/social/google");

        SocialAccessToken res = webClient.post()
            .headers(headers -> headers.add("Content-Type",
                "application/x-www-form-urlencoded;charset=utf-8"))
            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
            .body(BodyInserters.fromFormData(formData))
            .exchangeToMono(response -> response.bodyToMono(SocialAccessToken.class))
            .block();

        if (res == null || res.access_token() == null) {
            throw new CustomException(Error.FAIL_SOCIAL_LOGIN);
        }

        return res.access_token();
    }

    @Override
    public SocialUserInfo getUserInfo(String accessToken) {
        WebClient webClient = WebClient.create("https://www.googleapis.com/oauth2/v1/userinfo?alt=json");

        GoogleUserInfo googleUserInfo = webClient.get()
            .headers(header -> {
                header.add("Authorization", "Bearer " + accessToken);
            })
            .exchangeToMono(response -> response.bodyToMono(GoogleUserInfo.class))
            .block();

        if (googleUserInfo == null || googleUserInfo.id() == null) {
            throw new CustomException(Error.FAIL_SOCIAL_LOGIN);
        }

        return new SocialUserInfo(googleUserInfo.id(), googleUserInfo.email(), googleUserInfo.name(), AccountType.GOOGLE);
    }

    @Override
    SocialPlatform getSocialPlatform() {
        return SocialPlatform.GOOGLE;
    }

    record GoogleUserInfo(String id, String email, String name, String picture) {
    }
}

File: qc-core/qc-user/qc-user-social/src/main/kotlin/com/eager/questioncloud/social/google/GoogleAPI.kt
package com.eager.questioncloud.social.google

import com.eager.questioncloud.common.http.ContentType
import com.eager.questioncloud.common.http.HttpClient
import com.eager.questioncloud.common.http.HttpRequest
import com.eager.questioncloud.social.SocialAPI
import com.eager.questioncloud.social.SocialAccessToken
import com.eager.questioncloud.social.SocialPlatform
import com.eager.questioncloud.social.SocialUserInfo
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class GoogleAPI(
    private val httpClient: HttpClient
) : SocialAPI {
    @Value("\${GOOGLE_CLIENT_ID}")
    private lateinit var GOOGLE_CLIENT_ID: String
    
    @Value("\${GOOGLE_CLIENT_SECRET}")
    private lateinit var GOOGLE_CLIENT_SECRET: String
    
    @Value("\${CLIENT_URL}")
    private lateinit var CLIENT_URL: String
    
    override fun getAccessToken(code: String): String {
        val form = mutableMapOf<String, String>()
        form["client_id"] = GOOGLE_CLIENT_ID
        form["client_secret"] = GOOGLE_CLIENT_SECRET
        form["code"] = code
        form["grant_type"] = "authorization_code"
        form["redirect_uri"] = "$CLIENT_URL/user/social/kakao"
        
        val request = HttpRequest(url = "https://oauth2.googleapis.com/token", form = form, contentType = ContentType.FORM)
        val response = httpClient.post(request, SocialAccessToken::class.java)
        return response.access_token
    }
    
    override fun getUserInfo(accessToken: String): SocialUserInfo {
        val headers = mapOf("Authorization" to "Bearer $accessToken")
        val request = HttpRequest(url = "https://www.googleapis.com/oauth2/v1/userinfo?alt=json", headers = headers)
        val response = httpClient.get(request, GoogleUserInfo::class.java)
        return SocialUserInfo(response.id, response.email, response.name, SocialPlatform.GOOGLE)
    }
    
    override fun getSocialPlatform(): SocialPlatform {
        return SocialPlatform.GOOGLE
    }
}

private data class GoogleUserInfo(val id: String, val email: String, val name: String, val picture: String)
